lenskit-eval/src/main/java/org/lenskit/eval/traintest/recommend/TopNNDPMMetric.java

```diff
@@ -0,0 +1,175 @@
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collections;
+
+/**
+ * Measure the nDPM of the top-N recommendations, using rankings.
+ * This metric is registered with the type name `ndpm`.
+ * The paper used as a reference for this implementation is http://www2.cs.uregina.ca/~yyao/PAPERS/jasis_ndpm.pdf.
+ */
+public class TopNNDPMMetric extends TopNMetric<MeanAccumulator> {
+    public static final String DEFAULT_COLUMN = "TopN.nDPM";
+
+    /**
+     * Construct a top-N nDCG metric from a spec.
+     * @param spec The spec.
+     */
+    @JsonCreator
+    public TopNNDPMMetric(Spec spec) {this(spec.getColumnName());
+    }
+
+    /**
+     * Construct a new nDPM Top-N metric.
+     */
+    public TopNNDPMMetric(String name) {
+        super(Collections.singletonList(StringUtils.defaultString(name, DEFAULT_COLUMN)),
+              Collections.singletonList(StringUtils.defaultString(name, DEFAULT_COLUMN)));
+    }
+
+    @Nullable
+    @Override
+    public MeanAccumulator createContext(AlgorithmInstance algorithm, DataSet dataSet, org.lenskit.api.Recommender recommender) {
+        return new MeanAccumulator();
+    }
+
+    @Nonnull
+    @Override
+    public MetricResult getAggregateMeasurements(MeanAccumulator context) {
+        return MetricResult.singleton(DEFAULT_COLUMN, context.getMean());
+    }
+
+    @Nonnull
+    @Override
+    public MetricResult measureUser(TestUser user, int targetLength, ResultList recommendations, MeanAccumulator context) {
+        if (recommendations == null) {
+            return MetricResult.empty();
+        }
+
+        Long2DoubleMap ratings = user.getTestRatings();
+
+        long[] actual = LongUtils.asLongCollection(recommendations.idList()).toLongArray();
+
+        double dpm = computeDPM(actual, ratings);
+
+        double normalizingFactor = computeNormalizingFactor(actual, ratings);
+
+        double nDPM = dpm / normalizingFactor; // Normalized nDPM
+
+        context.add(nDPM);
+
+        return MetricResult.singleton(DEFAULT_COLUMN, nDPM);
+    }
+
+    /**
+     * Compute dpm of list of items, with respect to user's ratings.
+     */
+
+    double computeDPM(long [] actual_item, Long2DoubleFunction value) {
+        int nCompatible = 0;
+        int nDisagree = 0;
+
+        for(int i = 0; i < actual_item.length; i++){
+            for(int j = i+1; j < actual_item.length; j++){
+                double valueOne = 0;
+                double valueTwo = 0;
+
+                if (value.containsKey(actual_item[i])) {
+                    valueOne = value.get(actual_item[i]);
+
+                    if (value.containsKey(actual_item[j])) {
+                        valueTwo = value.get(actual_item[j]);
+                    }
+                }
+                if(valueOne < valueTwo){
+                    nDisagree++;
+                }
+                if(valueOne == valueTwo) {
+                    nCompatible++;
+                }
```

Comment:
Is it correct to count items as compatible if the user's ordering does not assert anything about them?